File: src/com/nerdscorner/mvp/mvp/busevents/base/MvpComponent.java
package com.nerdscorner.mvp.mvp.busevents.base;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

import com.nerdscorner.mvp.mvp.BaseComponent;
import com.nerdscorner.mvp.utils.FileCreator;

public abstract class MvpComponent extends BaseComponent {

    protected final String basePath;
    protected final String basePackage;
    protected final String screenName;

    private String componentTemplate;
    private String componentWithExtension;

    public MvpComponent(String basePath, String basePackage, String screenName, String componentTemplate, String componentWithExtension,
                        boolean shouldCreateWiring) {
        this.basePath = basePath;
        this.basePackage = basePackage;
        this.screenName = screenName;
        this.componentTemplate = getTemplatePath(shouldCreateWiring, componentTemplate);
        this.componentWithExtension = componentWithExtension;
    }

    private String getTemplatePath(boolean shouldCreateWiring, String componentTemplate) {
        return "/templates/" + (shouldCreateWiring ? "wiring/" : "clean/") + componentTemplate;
    }

    public boolean build(boolean isJava) {
        try {
            //Create {$screenName}{$Component}.{$languageExtension}
            componentTemplate = componentTemplate.replace(LANGUAGE_PLACEHOLDER, isJava ? JAVA_FOLDER : KOTLIN_FOLDER);
            componentWithExtension = componentWithExtension.replace(LANGUAGE_EXTENSION_PLACEHOLDER, isJava ? JAVA_EXTENSION : KOTLIN_EXTENSION);
            InputStream template = getClass().getResourceAsStream(componentTemplate);
            File component = new File(basePath + relativePath(), componentWithExtension);
            FileCreator.createFile(template, component, basePackage, screenName);
            return true;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    protected abstract String relativePath();

    @Override
    public void rollback() {
        new File(basePath, screenName + componentWithExtension).delete();
    }
}

File: src/com/nerdscorner/mvp/mvp/busevents/base/MvpComponent.kt
package com.nerdscorner.mvp.mvp.busevents.base

import com.nerdscorner.mvp.domain.ExecutionResult
import com.nerdscorner.mvp.mvp.BaseComponent
import com.nerdscorner.mvp.utils.FileCreator
import java.io.File

abstract class MvpComponent(
        protected val basePath: String,
        protected val basePackage: String,
        protected val screenName: String,
        componentTemplate: String,
        private var componentWithExtension: String,
        shouldCreateWiring: Boolean
) : BaseComponent() {

    private var componentTemplate: String = getTemplatePath(shouldCreateWiring, componentTemplate)

    private fun getTemplatePath(shouldCreateWiring: Boolean, componentTemplate: String): String {
        return "/templates/" + (if (shouldCreateWiring) "wiring/" else "clean/") + componentTemplate
    }

    fun build(isJava: Boolean): ExecutionResult {
        try {
            //Create {$screenName}{$Component}.{$languageExtension}
            componentTemplate = componentTemplate.replace(
                    LANGUAGE_PLACEHOLDER,
                    if (isJava) {
                        JAVA_FOLDER
                    } else {
                        KOTLIN_FOLDER
                    }
            )
            componentWithExtension = componentWithExtension.replace(
                    LANGUAGE_EXTENSION_PLACEHOLDER,
                    if (isJava) {
                        JAVA_EXTENSION
                    } else {
                        KOTLIN_EXTENSION
                    }
            )
            val template = javaClass.getResourceAsStream(componentTemplate) ?: return ExecutionResult(false)
            val component = File(absolutePath(), componentWithExtension)
            FileCreator.createFile(template, component, basePackage, screenName)
            return ExecutionResult(true)
        } catch (e: Exception) {
            e.printStackTrace()
            return ExecutionResult(false, e.message)
        }
    }

    protected abstract fun relativePath(): String

    private fun absolutePath() = basePath + relativePath()

    override fun rollback() {
        try {
            File(absolutePath(), componentWithExtension).delete()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
